File: server/src/main/java/avatar/Farm/Animal.java
package avatar.Farm;

public class
Animal {
    private int id;//id vat nuoi
    private int health;              // Sức khỏe của động vật
    private int level;               // Cấp độ của động vật
    private int resourceCount;       // Số tài nguyên (sản phẩm) mà động vật tạo ra
    private int nextProductionTime;  // Thời gian để động vật tạo ra sản phẩm tiếp theo
    private boolean isAlive;         // Trạng thái động vật có còn sống không
    private boolean isReadyForBreeding; // Trạng thái động vật sẵn sàng sinh sản không
    private boolean isHarvestable;   // Trạng thái động vật có thể thu hoạch sản phẩm không

    // Constructor
    public Animal(int id,int health, int level, int resourceCount, int nextProductionTime, boolean isAlive, boolean isReadyForBreeding, boolean isHarvestable) {
        this.id = id;
        this.health = health;
        this.level = level;
        this.resourceCount = resourceCount;
        this.nextProductionTime = nextProductionTime;
        this.isAlive = isAlive;
        this.isReadyForBreeding = isReadyForBreeding;
        this.isHarvestable = isHarvestable;
    }


    // Getters và Setters
    public int getId() {
        return id;
    }

    public void setId(int Id) {
        this.id = Id;
    }


    // Getters và Setters
    public int getHealth() {
        return health;
    }

    public void setHealth(int health) {
        this.health = health;
    }

    public int getLevel() {
        return level;
    }

    public void setLevel(int level) {
        this.level = level;
    }

    public int getResourceCount() {
        return resourceCount;
    }

    public void setResourceCount(int resourceCount) {
        this.resourceCount = resourceCount;
    }

    public int getNextProductionTime() {
        return nextProductionTime;
    }

    public void setNextProductionTime(int nextProductionTime) {
        this.nextProductionTime = nextProductionTime;
    }

    public boolean isAlive() {
        return isAlive;
    }

    public void setAlive(boolean alive) {
        isAlive = alive;
    }

    public boolean isReadyForBreeding() {
        return isReadyForBreeding;
    }

    public void setReadyForBreeding(boolean readyForBreeding) {
        isReadyForBreeding = readyForBreeding;
    }

    public boolean isHarvestable() {
        return isHarvestable;
    }

    public void setHarvestable(boolean harvestable) {
        isHarvestable = harvestable;
    }
    public byte getType() {
        // Ví dụ trả về một ID động vật ngẫu nhiên (hoặc có thể được xác định theo loại động vật cụ thể)
        return (byte) (50 + level % 7);  // Giả sử ID động vật phụ thuộc vào cấp độ
    }
    @Override
    public String toString() {
        return "Animal{" +
                "health=" + health +
                ", level=" + level +
                ", resourceCount=" + resourceCount +
                ", nextProductionTime=" + nextProductionTime +
                ", isAlive=" + isAlive +
                ", isReadyForBreeding=" + isReadyForBreeding +
                ", isHarvestable=" + isHarvestable +
                '}';
    }
}

File: server-node/src/farm/Animal.js
/**
 * Port của avatar/Farm/Animal.java.
 * NOTE: JS không cho vừa field `isAlive` vừa method `isAlive()` ⇒ các cờ boolean
 * lưu ở `_isAlive`/`_isReadyForBreeding`/`_isHarvestable`, truy cập qua method
 * đúng tên Java (isAlive(), setAlive(), ...).
 */
export class Animal {
  constructor(id = 0, health = 0, level = 0, resourceCount = 0, nextProductionTime = 0,
              isAlive = false, isReadyForBreeding = false, isHarvestable = false) {
    this.id = id | 0;                        // id vat nuoi
    this.health = health | 0;                // Sức khỏe của động vật
    this.level = level | 0;                  // Cấp độ của động vật
    this.resourceCount = resourceCount | 0;  // Số tài nguyên (sản phẩm) mà động vật tạo ra
    this.nextProductionTime = nextProductionTime | 0; // Thời gian tạo sản phẩm tiếp theo
    this._isAlive = !!isAlive;               // Trạng thái động vật có còn sống không
    this._isReadyForBreeding = !!isReadyForBreeding; // Sẵn sàng sinh sản không
    this._isHarvestable = !!isHarvestable;   // Có thể thu hoạch sản phẩm không
  }

  // Getters và Setters
  getId() { return this.id; }
  setId(Id) { this.id = Id | 0; }

  getHealth() { return this.health; }
  setHealth(health) { this.health = health | 0; }

  getLevel() { return this.level; }
  setLevel(level) { this.level = level | 0; }

  getResourceCount() { return this.resourceCount; }
  setResourceCount(resourceCount) { this.resourceCount = resourceCount | 0; }

  getNextProductionTime() { return this.nextProductionTime; }
  setNextProductionTime(nextProductionTime) { this.nextProductionTime = nextProductionTime | 0; }

  isAlive() { return this._isAlive; }
  setAlive(alive) { this._isAlive = !!alive; }

  isReadyForBreeding() { return this._isReadyForBreeding; }
  setReadyForBreeding(readyForBreeding) { this._isReadyForBreeding = !!readyForBreeding; }

  isHarvestable() { return this._isHarvestable; }
  setHarvestable(harvestable) { this._isHarvestable = !!harvestable; }

  getType() {
    // Giả sử ID động vật phụ thuộc vào cấp độ
    return ((50 + (this.level % 7)) << 24) >> 24; // (byte)
  }

  toString() {
    return 'Animal{'
      + 'health=' + this.health
      + ', level=' + this.level
      + ', resourceCount=' + this.resourceCount
      + ', nextProductionTime=' + this.nextProductionTime
      + ', isAlive=' + this._isAlive
      + ', isReadyForBreeding=' + this._isReadyForBreeding
      + ', isHarvestable=' + this._isHarvestable
      + '}';
  }
}

export default Animal;
